py-polars/polars/io/_utils.py

```diff
@@ -223,3 +230,100 @@ def _process_http_file(path: str, encoding: str | None = None) -> BytesIO:
             return BytesIO(f.read())
         else:
             return BytesIO(f.read().decode(encoding).encode("utf8"))
+
+
+@overload
+def _prepare_write_file_arg(
+    file: str | Path | BytesIO, **kwargs: Any
+) -> ContextManager[str | BytesIO]:
+    ...
+
+
+@overload
+def _prepare_write_file_arg(  # type: ignore[misc]
+    file: BinaryIO | BytesIO | str | Path, **kwargs: Any
+) -> ContextManager[str | BytesIO | BinaryIO]:
+    ...
```

Comment:
looks like `Path` and `BytesIO` appear in both of these overloads? I think that's why `mypy` is telling you that they overlap

also, I think you'll need to include
```
    encoding: str | None = ...,
    pyarrow_options: dict[str, Any] | None = ...
```
in the overloads

Reply:
I see, but I am not entirely sure how to solve that. Because if I change any of these overloads, then I will get a mypy error in frame.py on "file". Because the overload input matches each individual writer file input and expect output.

Reply:
@MarcoGorelli I think, I got it. Removing None as a possible type and return type from the main function removes the error.

Reply:
@MarcoGorelli  I've pushed a change that fixes the error. I think I still grossly misunderstood before on how to use it properly.

Reply:
nice, thanks!

I think you might be able to make this cleaner with `TypeVar`

example: you could write this with overloads
```python
from typing import TypeVar
from typing import overload
from datetime import date

@overload
def my_trivial_function(x: date) -> date:
    ...

@overload
def my_trivial_function(x: str) -> str:
    ...

@overload
def my_trivial_function(x: int) -> int:
    ...

def my_trivial_function(x: int | str | date) -> int | str | date:
    return x


variable: int
my_trivial_function(variable)  # builtins.int
```

but you could also use `TypeVar` to write it as:
```python
from typing import TypeVar
from datetime import date

T = TypeVar('T', int, str, date)

def my_trivial_function(x: T) -> T:
    return x


variable: int
my_trivial_function(variable)  # builtins.int
```

Reply:
I've implemented it with TypeVar, much cleaner. Thanks for the suggestion! I do have a question, is there still any use to the overload decorator?

Reply:
@MarcoGorelli  it seems if one uses TypeVar you can't use if TYPE_CHECKING: anymore for the imports. And a lazy import for xlsxwriter has not been implemented (`from polars.dependencies import xlsxwriter`). 

I am reverting the change and keeping the overloads : )

Reply:
> if one uses TypeVar you can't use if TYPE_CHECKING: anymore for the imports

sure, but can you define the `TypeVar` within the `TYPE_CHECKING` block too?

Reply:
@MarcoGorelli let me try that!

Reply:
Ah yes, that works! 😄

Reply:
nice one!

I haven't looked at the rest, but I'd also suggest choosing a more descriptive name - perhaps `FileType` instead of `T`?

Reply:
@MarcoGorelli  Done!